### PR TITLE
[Refactor] Add nodiscard and check status return

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 12 / 30 (40.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/table_function/generate_series.h|0|5|00.00%|[38, 43, 45, 47, 49]|
|:large_blue_circle:|src/exprs/table_function/table_function.h|0|1|00.00%|[56]|
|:large_blue_circle:|src/exec/stream/state/mem_state_table.h|0|1|00.00%|[39]|
|:large_blue_circle:|src/exprs/table_function/list_rowsets.h|0|5|00.00%|[37, 42, 44, 46, 48]|
|:large_blue_circle:|src/exprs/table_function/multi_unnest.h|0|5|00.00%|[99, 104, 106, 108, 110]|
|:large_blue_circle:|src/exprs/table_function/json_each.h|4|5|80.00%|[37]|
|:large_blue_circle:|src/util/timezone_hsscan.h|1|1|100.00%|[]|
|:large_blue_circle:|src/exprs/table_function/unnest.h|5|5|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.h|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 23 / 41 (56.10%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/table_function/generate_series.h|0|5|00.00%|[38, 43, 45, 46, 49]|
|:large_blue_circle:|src/exprs/table_function/table_function.h|0|1|00.00%|[56]|
|:large_blue_circle:|src/exec/stream/state/mem_state_table.h|0|1|00.00%|[39]|
|:large_blue_circle:|src/exprs/table_function/list_rowsets.h|0|5|00.00%|[37, 42, 44, 45, 48]|
|:large_blue_circle:|src/exprs/table_function/multi_unnest.h|0|5|00.00%|[99, 104, 106, 107, 110]|
|:large_blue_circle:|src/exprs/table_function/json_each.h|4|5|80.00%|[37]|
|:large_blue_circle:|src/util/timezone_hsscan.h|1|1|100.00%|[]|
|:large_blue_circle:|src/exprs/table_function/unnest.h|5|5|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exprs/time_functions.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.cpp|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 23 / 41 (56.10%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/table_function/generate_series.h|0|5|00.00%|[38, 43, 45, 46, 49]|
|:large_blue_circle:|src/exprs/table_function/table_function.h|0|1|00.00%|[56]|
|:large_blue_circle:|src/exec/stream/state/mem_state_table.h|0|1|00.00%|[39]|
|:large_blue_circle:|src/exprs/table_function/list_rowsets.h|0|5|00.00%|[37, 42, 44, 45, 48]|
|:large_blue_circle:|src/exprs/table_function/multi_unnest.h|0|5|00.00%|[99, 104, 106, 107, 110]|
|:large_blue_circle:|src/exprs/table_function/json_each.h|4|5|80.00%|[37]|
|:large_blue_circle:|src/util/timezone_hsscan.h|1|1|100.00%|[]|
|:large_blue_circle:|src/exprs/table_function/unnest.h|5|5|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exprs/time_functions.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.cpp|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 23 / 41 (56.10%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/table_function/generate_series.h|0|5|00.00%|[38, 43, 45, 46, 49]|
|:large_blue_circle:|src/exprs/table_function/table_function.h|0|1|00.00%|[56]|
|:large_blue_circle:|src/exec/stream/state/mem_state_table.h|0|1|00.00%|[39]|
|:large_blue_circle:|src/exprs/table_function/list_rowsets.h|0|5|00.00%|[37, 42, 44, 45, 48]|
|:large_blue_circle:|src/exprs/table_function/multi_unnest.h|0|5|00.00%|[99, 104, 106, 107, 110]|
|:large_blue_circle:|src/exprs/table_function/json_each.h|4|5|80.00%|[37]|
|:large_blue_circle:|src/util/timezone_hsscan.h|1|1|100.00%|[]|
|:large_blue_circle:|src/exprs/table_function/unnest.h|5|5|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exprs/time_functions.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.cpp|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 29 / 52 (55.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/table_function/generate_series.h|0|5|00.00%|[38, 43, 45, 46, 49]|
|:large_blue_circle:|src/exprs/table_function/table_function.h|0|1|00.00%|[56]|
|:large_blue_circle:|src/exec/stream/state/mem_state_table.h|0|1|00.00%|[39]|
|:large_blue_circle:|src/exprs/table_function/list_rowsets.h|0|5|00.00%|[37, 42, 44, 45, 48]|
|:large_blue_circle:|src/exprs/table_function/multi_unnest.h|0|5|00.00%|[99, 104, 106, 107, 110]|
|:large_blue_circle:|src/exec/stream/state/mem_state_table.cpp|6|11|54.55%|[46, 54, 84, 129, 234]|
|:large_blue_circle:|src/exprs/table_function/json_each.h|4|5|80.00%|[37]|
|:large_blue_circle:|src/util/timezone_hsscan.h|1|1|100.00%|[]|
|:large_blue_circle:|src/exprs/table_function/unnest.h|5|5|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exprs/time_functions.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/stream/scan/stream_scan_operator.cpp|10|10|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

File: be/src/exprs/table_function/table_function.h [hunk 1]
```diff
@@ -79,18 +79,18 @@ class TableFunction {
...
 
-    virtual Status open(RuntimeState* runtime_state, TableFunctionState* state) const = 0;
+    [[nodiscard]] virtual Status open(RuntimeState* runtime_state, TableFunctionState* state) const = 0;
 
     //Table function processing logic
     virtual std::pair<Columns, UInt32Column::Ptr> process(TableFunctionState* state) const = 0;
 
     //Release the resources constructed in init and prepare
-    virtual Status close(RuntimeState* runtime_state, TableFunctionState* context) const = 0;
+    [[nodiscard]] virtual Status close(RuntimeState* runtime_state, TableFunctionState* context) const = 0;
```
Comment (silverbullet233) [inline be/src/exprs/table_function/table_function.h:93]: can we change this return type to void? since all override implementation always return OK.

Comment (LiShuMing) [reply be/src/exprs/table_function/table_function.h:93]: It's better to keep it. For now, `JavaUDTFFunction` will use it:
```
Status JavaUDTFFunction::close(RuntimeState* runtime_state, TableFunctionState* state) const {
    auto promise = call_function_in_pthread(runtime_state, [state]() {
        delete state;
        return Status::OK();
    });
    RETURN_IF_ERROR(promise->get_future().get());
    return Status::OK();
}
```

And it's better to keep it for the extension?

**Review by silverbullet233 — APPROVED** [review]

**Pull request merged**